### PR TITLE
[MERGE] stock,mrp,purchase,sale: rescheduling propagation 

### PR DESCRIPTION
**Pull request by Whenrow**
Currently the fact that the changes in the expected dates are propagated
or not is based on :
- the value of the field propagate on the stock.rule 
- the general inventory setting setting use_propagation_minimum_delta

That's not easy to understand for users and difficult to configure. 

The purpose of this task is to clarify this configuration (everything is configured
at the same place + but we keep the flexibility by setting this config on the stock.rule) 

task-1913401

### CODE REVIEW

File: addons/sale_stock/tests/test_sale_stock_lead_time.py [hunk 1]
```diff
@@ -98,3 +98,163 @@ def test_01_product_route_level_delays(self):
...
+
+        # Confirm our standard sale order
+        order.action_confirm()
+
+        # Check the picking crated or not
+        self.assertTrue(order.picking_ids, "Pickings should be created.")
+
+        # Check schedule date of ship type picking
+        out = order.picking_ids.filtered(lambda r: r.picking_type_id == self.warehouse_1.out_type_id)
+        out_min_date = fields.Datetime.from_string(out.scheduled_date)
```
Comment (sle-odoo) [inline addons/sale_stock/tests/test_sale_stock_lead_time.py]: remove from string

File: addons/sale_stock/tests/test_sale_stock_lead_time.py [hunk 2]
```diff
@@ -98,3 +98,163 @@ def test_01_product_route_level_delays(self):
...
+        out_date = fields.Datetime.from_string(order.date_order) + timedelta(days=self.product_1.sale_delay) - timedelta(days=out.move_lines[0].rule_id.delay)
+        self.assertTrue(abs(out_min_date - out_date) <= timedelta(seconds=1), 'Schedule date of ship type picking should be equal to: order date + Customer Lead Time - pull rule delay.')
```
Comment (sle-odoo) [inline addons/sale_stock/tests/test_sale_stock_lead_time.py]: i'm not sure we should subsctract - pull rule delay, we need to check with lap

File: addons/sale_stock/tests/test_sale_stock_lead_time.py [hunk 3]
```diff
@@ -98,3 +98,163 @@ def test_01_product_route_level_delays(self):
...
+
+        # Now change the schedule date of pick
+        # Note : pack and out has change scheduled_date automatically based on delay set on pick
+        pick.write({'scheduled_date': pick_min_date + timedelta(days=5)})
+
+        # Now check scheduled_date of pack and out are changed or not based on propagate is true on rules?
+        self.assertEquals(pack.scheduled_date, (pack_min_date + timedelta(days=5)),
+            'Schedule date of pack should be changed based on delay.')
+        self.assertEquals(out.scheduled_date, (out_min_date + timedelta(days=5)),
+            'Schedule date of out should be changed based on delay.')
```
Comment (sle-odoo) [inline addons/sale_stock/tests/test_sale_stock_lead_time.py]: complete this test, simulate _action_done on the pack by writing on on it date + timedeldtay 10 and state done, check the out_date is incremented by 10 days

File: addons/stock/views/stock_rule_views.xml [hunk 4]
```diff
@@ -103,6 +103,13 @@
                                 <field name="partner_address_id" attrs="{'invisible': [('action', '=', 'push')]}"/>
                                 <field name="delay"/>
                             </group>
+                            <group name="scheduling_group" string="Scheduling" attrs="{'invisible': [('action', '=', 'push')]}" groups="base.group_no_one">
+                                <field name="propagate_date"/>
+                                <label for="propagate_date_minimum_delta" attrs="{'invisible': [('propagate_date', '=', False)]}"/>
+                                <div attrs="{'invisible': [('propagate_date', '=', False)]}">
+                                    <field name="propagate_date_minimum_delta" class="oe_inline"/> day(s)
+                                </div>
+                            </group>
```
Comment (sle-odoo) [inline addons/stock/views/stock_rule_views.xml]: add a basic test: create two chained moves, enable propate + minimum delta, change 

with two chained moves
1) enable propagate, change expected dat
2) enabe propagate, action_done on the first
3) disable propagate, change expected dat
4) disable propagate, action_done on the first

File: addons/sale_stock/tests/test_sale_stock_lead_time.py (hunk 3 above)
Comment (sle-odoo) [reply addons/sale_stock/tests/test_sale_stock_lead_time.py]: actually no, just add the tests i said in stock

add a test showing that po lines aren't merged if they have different expected dates

File: addons/mrp/models/stock_rule.py [hunk 5]
```diff
@@ -87,6 +87,8 @@ def _prepare_mo_vals(self, product_id, product_qty, product_uom, location_id, na
             'date_planned_finished': values['date_planned'],
             'procurement_group_id': False,
             'propagate': self.propagate,
+            'propagate_date': self.propagate_date,
+            'propagate_date_minimum_delta': self.propagate_date_minimum_delta,
```
Comment (sle-odoo) [inline addons/mrp/models/stock_rule.py]: add a test

File: addons/purchase_stock/models/purchase.py [hunk 6]
```diff
@@ -266,12 +268,18 @@ def create(self, values):
 
     @api.multi
     def write(self, values):
+        for line in self:
+            if values.get('date_planned') and line.propagate_date:
+                new_date = fields.Datetime.from_string(values['date_planned'])
```
Comment (amoyaux) [inline addons/purchase_stock/models/purchase.py]: values['date_planned'] is already a datetime no?

File: addons/stock/models/stock_move.py [hunk 7]
```diff
@@ -412,24 +414,17 @@ def write(self, vals):
...
-                        if abs(delta_days) >= move.company_id.propagation_minimum_delta:
-                            old_move_date = move.move_dest_ids[0].date_expected
-                            new_move_date = (old_move_date + relativedelta.relativedelta(days=delta_days or 0)).strftime(DEFAULT_SERVER_DATETIME_FORMAT)
-                            propagated_changes_dict['date_expected'] = new_move_date
-                    #For pushed moves as well as for pulled moves, propagate by recursive call of write().
-                    #Note that, for pulled moves we intentionally don't propagate on the procurement.
-                    if propagated_changes_dict:
-                        move.move_dest_ids.filtered(lambda m: m.state not in ('done', 'cancel')).write(propagated_changes_dict)
+                if move.move_dest_ids and move.propagate_date:
+                    new_date = fields.Datetime.from_string(vals.get(propagated_date_field))
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py]: remove fields.Datetime.from_string

**Comment by sle-odoo** [comment]
robodoo r+ rebase-merge

**Pull request closed without merging**